src/h5web/visualizations/nexus/utils.test.ts

```diff
@@ -0,0 +1,330 @@
+    });
+
+    it('should throw if default entity is not found', () => {
+      const group = makeGroup('foo', [], {
+        attributes: [makeStrAttr('default', 'bar')],
+      });
+
+      expect(() => findNxDataGroup(group)).toThrow(/entity at path/u);
+    });
+
+    it('should throw if default entity is not a group', () => {
+      const group = makeGroup('foo', [makeDatatype('bar', intType)], {
+        attributes: [makeStrAttr('default', 'bar')],
+      });
+
+      expect(() => findNxDataGroup(group)).toThrow(/group at path/u);
+    });
+
+    it('should throw if default group is not `NXdata` and has no `default` attribute', () => {
+      const group = makeGroup('foo', [makeGroup('bar')], {
+        attributes: [makeStrAttr('default', 'bar')],
+      });
+
+      expect(() => findNxDataGroup(group)).toThrow(/NXdata group/u);
+    });
+  });
+
+  describe('findSignalDataset', () => {
+    it("should return dataset named 'signal'", () => {
+      const dataset = makeDataset('dataset', makeSimpleShape([1]), intType);
+      const group = makeGroup('group', [dataset], {
+        attributes: [makeStrAttr('signal', 'dataset')],
+      });
+
+      expect(findSignalDataset(group)).toBe(dataset);
+    });
+
+    it("should throw if group doesn't have 'signal' attribute", () => {
+      const group = makeGroup('group');
+      expect(() => findSignalDataset(group)).toThrow(/'signal' attribute$/u);
+    });
+
+    it("should throw if group has 'signal' attribute with wrong type", () => {
+      const group = makeGroup('group', [], {
+        attributes: [makeScalarIntAttr('signal', 42)],
+      });
+
+      expect(() => findSignalDataset(group)).toThrow(/to be a string/u);
+    });
+
+    it("should throw if signal entity doesn't exist", () => {
+      const group = makeGroup('group', [], {
+        attributes: [makeStrAttr('signal', 'foo')],
+      });
+
+      expect(() => findSignalDataset(group)).toThrow(/to exist/u);
+    });
+
+    it('should throw if signal entity is not a dataset', () => {
+      const group = makeGroup('group', [makeGroup('foo')], {
+        attributes: [makeStrAttr('signal', 'foo')],
+      });
+
+      expect(() => findSignalDataset(group)).toThrow(/to be a dataset/u);
+    });
+
+    it('should throw if signal dataset has non-simple shape', () => {
+      const dataset = makeDataset('dataset', scalarShape, intType);
+      const group = makeGroup('group', [dataset], {
+        attributes: [makeStrAttr('signal', 'dataset')],
+      });
+
+      expect(() => findSignalDataset(group)).toThrow(/to have simple shape/u);
+    });
+
+    it('should throw if signal dataset has no dimensions', () => {
+      const dataset = makeDataset('dataset', makeSimpleShape([]), stringType);
+      const group = makeGroup('group', [dataset], {
+        attributes: [makeStrAttr('signal', 'dataset')],
+      });
+
+      expect(() => findSignalDataset(group)).toThrow(/to have dimensions/u);
+    });
```

Comment:
Actually, I am not sure this case is possible. Either a dataset has a scalar shape or it has a simple shape and should have dims in this case.

Doesn't hurt to keep it, though.

Reply:
This checks for a shape with `dims` equal to `[]`, which makes no sense.

Reply:
I mean you'd hope that a back-end would never return this, so perhaps it is indeed useless to assert it.